provider/resource.go

```diff
@@ -235,6 +237,60 @@ func detectCloudShell() cloudShellProfile {
 	return negative
 }
 
+// stringValue gets a string value from a property map if present, else ""
+func stringValue(vars resource.PropertyMap, prop resource.PropertyKey, envs []string) string {
+	val, ok := vars[prop]
+	if ok && val.IsString() {
+		return val.StringValue()
+	}
+	for _, env := range envs {
+		val, ok := os.LookupEnv(env)
+		if ok {
+			return val
+		}
+	}
+	return ""
+}
+
+// preConfigureCallback returns an error when cloud provider setup is misconfigured
+func preConfigureCallback(vars resource.PropertyMap, c tfshim.ResourceConfig) error {
+
+	envName := stringValue(vars, "environment", []string{"ARM_ENVIRONMENT"})
+	if envName == "" {
+		envName = "public"
+	}
+
+	// validate the azure config
+	// make a Builder
+	builder := &authentication.Builder{
+		SubscriptionID:     stringValue(vars, "subscriptionID", []string{"ARM_SUBSCRIPTION_ID"}),
+		ClientID:           stringValue(vars, "clientId", []string{"ARM_CLIENT_ID"}),
+		ClientSecret:       stringValue(vars, "clientSecret", []string{"ARM_CLIENT_SECRET"}),
+		TenantID:           stringValue(vars, "tenantId", []string{"ARM_TENANT_ID"}),
+		Environment:        envName,
+		ClientCertPath:     stringValue(vars, "clientCertificatePath", []string{"ARM_CLIENT_CERTIFICATE_PATH"}),
+		ClientCertPassword: stringValue(vars, "clientCertificatePassword", []string{"ARM_CLIENT_CERTIFICATE_PASSWORD"}),
+		MsiEndpoint:        stringValue(vars, "msiEndpoint", []string{"ARM_MSI_ENDPOINT"}),
+		//AuxiliaryTenantIDs:   auxTenants,
+		ClientSecretDocsLink: "https://www.pulumi.com/docs/intro/cloud-providers/azure/setup/#service-principal-authentication",
+
+		// Feature Toggles
+		SupportsClientCertAuth:   true,
+		SupportsClientSecretAuth: true,
+		//SupportsManagedServiceIdentity: useMsi,
+		SupportsAzureCliToken: true,
+		//SupportsAuxiliaryTenants:       len(auxTenants) > 0,
+	}
+
+	_, err := builder.Build()
+
+	if err != nil {
+		return fmt.Errorf("failed to load application credentials. Please run 'az login' to set up account.\n" +
```

Comment:
I suggest wording it a little differently. In Azure, you authenticate as an application *or* a user. `az login` is only correct for the latter. Maybe

> failed to authenticate to Azure. Please sign in via `az login` or configure another authentication method. See...